File: spring-challenge.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <queue>
#include <tuple>

using namespace std;

struct Position
{
    int x;
    int y;

    bool operator == (Position p) {
        return (this->x == p.x) && (this->y == p.y);
    }
};

struct Pac
{
    int pacId;           // pac number (unique within a team)
    Position pos;        // position of pac in grid
    string typeId;       // unused in wood leagues
    int speedTurnsLeft;  // unused in wood leagues
    int abilityCooldown; // unused in wood leagues

    bool isAbilityAvailable() {
        return abilityCooldown == 0;
    }
};

struct Pellet
{
    Position pos; // pellet position in grid
    int value;    // amount of points this pellet is worth

    bool isSuper() {
        return value == 10;
    }
};

// TODO: Universal Grid cannot yet differentiate between WALL and EMPTY, maybe merge it with GRID in future
class GameState
{
public:
    int width, height;
    int myScore, opponentScore;
    vector<vector<char>> grid;
    vector<vector<int>> universalGrid;
    vector<Pac> pacs, enemies;
    vector<Pellet> pellets;

    GameState()
    {
        initGrid();
    }

    void initGrid()
    {
        cin >> width >> height;
        cin.ignore();

        vector<vector<char>> tempGrid(height, vector<char>(width, ' '));

        for (int i = 0; i < height; i++)
        {
            string row;
            getline(cin, row);
            for (int j = 0; j < width; j++)
            {
                tempGrid[i][j] = row[j];
            }
        }

        this->grid = tempGrid;

        vector<vector<int>> tempUniversalGrid(height, vector<int>(width, 0));
        for (int i = 0; i < height; i++)
        {
            for (int j = 0; j < width; j++)
            {
                if (grid[i][j] == ' ')
                    tempUniversalGrid[i][j] = 1;
                // else if (grid[i][j] == '#')
                //     tempGrid[i][j] = 0;
            }
        }
        this->universalGrid = tempUniversalGrid;
    }

    void removeStateFromGrid()
    {
        for (int i = 0; i < grid.size(); i++)
        {
            for (int j = 0; j < grid[i].size(); j++)
            {
                if (grid[i][j] != '#')
                    grid[i][j] = ' ';
            }
        }
    }

    void initUniversalGrid()
    {
        vector<vector<int>> tempGrid(height, vector<int>(width, 0));

        auto initIndex = [](int posx, int posy, auto &tempGrid, auto &grid) {
            if (grid[posy][posx] == ' ')
                tempGrid[posy][posx] = 0;
            else if (grid[posy][posx] == 'o')
                tempGrid[posy][posx] = 2;
            else if (grid[posy][posx] == 'O')
                tempGrid[posy][posx] = 10;
        };

        for (Pac pac : pacs)
        {
            for (int x = pac.pos.x; x >= 0 && grid[pac.pos.y][x] != '#'; x--)
                initIndex(x, pac.pos.y, tempGrid, grid);
            for (int y = pac.pos.y; y >= 0 && grid[y][pac.pos.x] != '#'; y--)
                initIndex(pac.pos.x, y, tempGrid, grid);
            for (int x = pac.pos.x; x < width && grid[pac.pos.y][x] != '#'; x++)
                initIndex(x, pac.pos.y, tempGrid, grid);
            for (int y = pac.pos.y; y < height && grid[y][pac.pos.x] != '#'; y++)
                initIndex(pac.pos.x, y, tempGrid, grid);
        }

        for (Pellet pellet : pellets)
        {
            tempGrid[pellet.pos.y][pellet.pos.x] = pellet.value == 10 ? 10 : 2;
        }

        universalGrid = tempGrid;

        // cerr << "universal grid initialised...\n";
    }

    void initGameState()
    {
        pacs.clear();
        enemies.clear();
        pellets.clear();
        removeStateFromGrid();

        cin >> myScore >> opponentScore;
        cin.ignore();

        int visiblePacCount;
        cin >> visiblePacCount;
        cin.ignore();

        // cerr << "basic input done...\n";

        for (int i = 0; i < visiblePacCount; i++)
        {
            int pacId, x, y, speedTurnsLeft, abilityCooldown;
            bool mine;
            string typeId;
            cin >> pacId >> mine >> x >> y >> typeId >> speedTurnsLeft >> abilityCooldown;
            cin.ignore();

            if (mine)
            {
                pacs.push_back(Pac{pacId, {x, y}, typeId, speedTurnsLeft, abilityCooldown});
            }
            else
            {
                enemies.push_back(Pac{pacId, {x, y}, typeId, speedTurnsLeft, abilityCooldown});
            }

            if (mine)
                this->grid[y][x] = 'P';
            else
                this->grid[y][x] = 'E';
        }

        // cerr << "pacs input done...\n";

        int visiblePelletCount;
        cin >> visiblePelletCount;
        cin.ignore();
        for (int i = 0; i < visiblePelletCount; i++)
        {
            int x, y, value;
            cin >> x >> y >> value;
            cin.ignore();

            pellets.push_back(Pellet{{x, y}, value});

            this->grid[y][x] = value == 10 ? 'O' : 'o';
        }

        // cerr << "pellets input done...\n";

        this->initUniversalGrid();
    }
};

class GameAI
{
private:
    GameState *gameState;

public:
    GameAI(GameState *_gameState)
    {
        gameState = _gameState;
    }

    void debugGrid()
    {
        auto grid = gameState->grid;
        for (auto row : grid)
        {
            for (auto ch : row)
            {
                cerr << ch;
            }
            cerr << endl;
        }
    }

    void debugUniversalGrid()
    {
        // while debugging show super pellets as 9 instead of 10
        auto grid = gameState->universalGrid;
        for (auto row : grid)
        {
            for (int ch : row)
            {
                if (ch == 10) ch = 9;
                cerr << ch << " ";
            }
            cerr << endl;
        }
    }

    Position getBestMovePosition(Pac pac, vector<vector<int>> &universalGridCopy) {
        const Position pos = pac.pos;
        const int width = this->gameState->width;
        const int height = this->gameState->height;

        const auto grid = this->gameState->grid;
        auto universalGrid = universalGridCopy;
        vector<vector<int>> scores(height, vector<int>(width, 0));
        vector<vector<bool>> visited(height, vector<bool>(width, false));

        auto getEntity = [=](Position pos) -> char {
            return grid[pos.y][pos.x];
        };

        auto setScore = [&](Position pos, int value) {
            if (scores[pos.y][pos.x] < value) scores[pos.y][pos.x] = value;
        };

        auto setVisited = [&](Position pos) {
            visited[pos.y][pos.x] = true;
        };

        Position superBestPos;

        // BFS lambda to encapsulate logic
        function<void(Position)> BFS = [&](Position curPos) {
            // if (depth > 15) return 1;
            queue<tuple<Position,int,int>> Q; // position, score, depth
            Q.push({ curPos, 0, 0 });

            while (!Q.empty()) {
                int size = Q.size();
                while (size--) {
                    auto [pos, score, depth] = Q.front();
                    Q.pop();

                    if (visited[pos.y][pos.x]) continue;

                    // check if wall then return
                    if (getEntity(pos) == '#') {
                        setVisited(pos);
                        continue;
                    }

                    // if (pos.x < 0 || pos.x >= width || pos.y < 0 || pos.y >= height) continue;

                    // initialise all direction positions correctly
                    Position tPos, bPos, rPos, lPos;

                    if (pos.x == 0) lPos = Position{width - 1, pos.y};
                    else lPos = Position{pos.x - 1, pos.y};

                    if (pos.x == width - 1) rPos = Position{0, pos.y};
                    else rPos = Position{pos.x + 1, pos.y};

                    if (pos.y == 0) tPos = Position{pos.x, height - 1};
                    else tPos = Position{pos.x, pos.y - 1};

                    if (pos.y == height - 1) bPos = Position{pos.x, 0};
                    else bPos = Position{pos.x, pos.y + 1};

                    // TODO: think more on this: adding -1 as a cost to walk
                    int newVal = score + universalGrid[pos.y][pos.x];
                    setScore(pos, newVal);
                    universalGrid[pos.y][pos.x] = -1000;

                    Q.push({ tPos, newVal, depth + 1 });
                    Q.push({ bPos, newVal, depth + 1 });
                    Q.push({ lPos, newVal, depth + 1 });
                    Q.push({ rPos, newVal, depth + 1 });

                    setVisited(pos);
                }
            }

        };

        BFS(pac.pos);

        int bestScore = 0;
        Position bestPos = pac.pos;
        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                int score = scores[i][j];
                // cerr << score << " ";

                if (score > bestScore) {
                    bestScore = score;
                    bestPos = Position{j,i};
                }
            }
            // cerr << endl;
        }

        // cerr << "id: " << pac.pacId << " score: " << bestScore << endl;
        // cerr << bestPos.x << " " << bestPos.y << endl << endl;

        return bestPos;
    }

    void nextMove()
    {
        vector<Pac> &pacs = gameState->pacs;
        vector<Pellet> &pellets = gameState->pellets;

        vector<vector<int>> universalGridCopy = this->gameState->universalGrid;

        vector<Position> targets;
        for (int i = 0; i < pacs.size(); i++)
        {
            Pac &pac = pacs[i];

            if (false && i > 0) {
                // for debugging
                cout << "MOVE " << pac.pacId << " " << pac.pos.x << " " << pac.pos.y;
            } else {
                // if SPEED ability available, just use it
                if (pac.isAbilityAvailable()) {
                    cout << "SPEED " << pac.pacId;
                } else {
                    Position pos = this->getBestMovePosition(pac, universalGridCopy);
                    if (find(targets.begin(), targets.end(), pos) != targets.end()) {
                        cout << "MOVE " << pac.pacId << " " << pac.pos.x << " " << pac.pos.y;
                    } else {
                        cout << "MOVE " << pac.pacId << " " << pos.x << " " << pos.y;
                    }
                    targets.push_back(pos);
                }
            }


            if (i != pacs.size() - 1)
                cout << "|";
        }
        cout << endl;
    }
};

int main()
{

    GameState gameState = GameState();
    GameAI ai{&gameState};

    while (1)
    {
        gameState.initGameState();
        // ai.debugUniversalGrid();
        ai.nextMove();
    }
}

File: spring-challenge.ts
// const readline = () => '';

const Entities = {
	EMPTY: ' ',
	WALL: '#',
	PAC: 'P',
	ENEMY: 'E',
	PELLET: 'o',
	SUPER_PELLET: 'O',
};

const Entity_Score = {
	[Entities.EMPTY]: -1,
	[Entities.WALL]: -1000,
	[Entities.PAC]: -10,
	[Entities.ENEMY]: -10,
	[Entities.PELLET]: 1,
	[Entities.SUPER_PELLET]: 10,
};

class Position {
	x: number;
	y: number;

	constructor(x: number, y: number) {
		this.x = x;
		this.y = y;
	}

	isEqual(pos: Position) {
		return this.x === pos.x && this.y === pos.y;
	}
};

class Pac {
	id: number;
	pos: Position;
	typeId: string;
	speedTurnsLeft: number;
	abilityCooldown: number;
	isDead: boolean;
	target: { pos: Position, score: number };

	constructor(id: number, pos: Position, _typeId: string, _speedTurnsLeft: number, _abilityCooldown: number) {
		this.id = id;
		this.pos = pos;
		this.typeId = _typeId;
		this.speedTurnsLeft = _speedTurnsLeft;
		this.abilityCooldown = _abilityCooldown;
		this.isDead = false;
	}

};

class Pellet {
	pos: Position;
	value: number;

	constructor(pos: Position, value: number) {
		this.pos = pos;
		this.value = value;
	}

	isSuper() {
		return this.value === 10;
	}
};

class Grid<T> {
	width: number;
	height: number;
	grid: T[][];

	constructor(width: number, height: number, cellDefaultValue?: any) {
		this.width = width;
		this.height = height;
		this.grid = [];
		for (let y = 0; y < height; y++) {
			this.grid.push([]);
			for (let x = 0; x < width; x++) {
				this.grid[y].push(cellDefaultValue);
			}
		}
	}

	get(pos: Position): T {
		return this.grid[pos.y][pos.x];
	}

	set(pos: Position, value: T) {
		this.grid[pos.y][pos.x] = value;
	}

	setRow(rowIndex: number, row: T[]) {
		this.grid[rowIndex] = row;
	}

	checkBounds(pos: Position) {
		const { width, height } = this;
		return !(pos.x < 0 || pos.x >= width || pos.y < 0 || pos.y >= height);
	}

	getNeighbours(pos: Position) {
		const neighbours: Position[] = [];
		let tPos: Position, bPos: Position, rPos: Position, lPos: Position;

		if (pos.x === 0) lPos = new Position(this.width - 1, pos.y);
		else lPos = new Position(pos.x - 1, pos.y);

		if (pos.x === this.width - 1) rPos = new Position(0, pos.y);
		else rPos = new Position(pos.x + 1, pos.y);

		if (pos.y === 0) tPos = new Position(pos.x, this.height - 1);
		else tPos = new Position(pos.x, pos.y - 1);

		if (pos.y === this.height - 1) bPos = new Position(pos.x, 0);
		else bPos = new Position(pos.x, pos.y + 1);

		neighbours.push(tPos, bPos, lPos, rPos);
		return neighbours;
	}

	debug() {
		for (let row of this.grid) {
			let outputStr = '';
			for (let col of row) {
				let colStr = String(col);
				outputStr += colStr;

				if (colStr.length === 2) outputStr += ' ';
				else outputStr += ' ';
			}
			console.error(outputStr);
		}
		console.error('\n\n');
	}
};

class GameState {
	grid: Grid<string>;
	universalGrid: Grid<number>;

	myScore: number;
	opponentScore: number;

	pacs = [] as Pac[];
	enemies = [] as Pac[];
	pellets = [] as Pellet[];

	constructor() {
		this.initGrid();
	}

	initGrid() {
		const inputs: string[] = readline().split(' ');
		const width = parseInt(inputs[0]);
		const height = parseInt(inputs[1]);

		this.grid = new Grid<string>(width, height, '');
		this.universalGrid = new Grid<number>(width, height, 1);

		for (let i = 0; i < height; i++) {
			const row: string = readline();
			this.grid.setRow(i, row.split(''));
		}
	}

	resetGridState() {
		for (let y = 0; y < this.grid.height; y++) {
			for (let x = 0; x < this.grid.width; x++) {
				if (this.grid.get(new Position(x, y)) !== Entities.WALL) {
					this.grid.set(new Position(x, y), Entities.EMPTY);
				} else {
					this.universalGrid.set(new Position(x, y), -10);
				}
			}
		}
	}

	initGameState() {
		this.resetGridState();

		const pacInput = [];
		const pelletInput = [];

		const initialInputs: string[] = readline().split(' ');
		this.myScore = parseInt(initialInputs[0]);
		this.opponentScore = parseInt(initialInputs[1]);

		const visiblePacCount: number = parseInt(readline());
		for (let i = 0; i < visiblePacCount; i++) {
			const inputs: string[] = readline().split(' ');
			const pacId: number = parseInt(inputs[0]);
			const mine: boolean = inputs[1] !== '0';
			const x: number = parseInt(inputs[2]);
			const y: number = parseInt(inputs[3]);
			const typeId: string = inputs[4];
			const speedTurnsLeft: number = parseInt(inputs[5]);
			const abilityCooldown: number = parseInt(inputs[6]);

			pacInput.push({ pacId, mine, x, y, typeId, speedTurnsLeft, abilityCooldown });
		}

		const visiblePelletCount: number = parseInt(readline());
		for (let i = 0; i < visiblePelletCount; i++) {
			const inputs: string[] = readline().split(' ');
			const x: number = parseInt(inputs[0]);
			const y: number = parseInt(inputs[1]);
			const value: number = parseInt(inputs[2]);

			pelletInput.push({ x, y, value });
			this.grid.set(new Position(x, y), value === 10 ? 'O' : 'o');
			this.universalGrid.set(new Position(x, y), value);
		}

		const createOrUpdatePac = (pacArray: Array<Pac>, { pacId, x, y, typeId, speedTurnsLeft, abilityCooldown }) => {
			const pacInstance = pacArray.find(pac => pac.id === pacId);
			if (pacInstance) {
				pacInstance.typeId = typeId;
				pacInstance.pos = new Position(x, y);
				pacInstance.speedTurnsLeft = speedTurnsLeft;
				pacInstance.abilityCooldown = abilityCooldown;
			} else {
				pacArray.push(new Pac(pacId, new Position(x, y), typeId, speedTurnsLeft, abilityCooldown));
			}
		};

		pacInput.forEach(({ mine, ...pacInfo }) => {
			if (mine) {
				createOrUpdatePac(this.pacs, pacInfo);
			} else {
				createOrUpdatePac(this.enemies, pacInfo);
			}

			this.grid.set(new Position(pacInfo.x, pacInfo.y), mine ? 'P' : 'E');
			this.universalGrid.set(new Position(pacInfo.x, pacInfo.y), -2);
		});

		const updateDeadStatus = (pacArray: Array<Pac>, pacInput: any[]) => {
			if (pacArray.length === pacInput.length) return;
			pacArray.forEach(pac => {
				const found = pacInput.filter(p => p.pacId === pac.id).length > 0;
				if (!found) {
					pac.isDead = true;
					// console.error(`${pac.id} not found in input, turning dead.`);
				}
			});
		};

		updateDeadStatus(this.pacs, pacInput.filter(p => p.mine));
		// updateDeadStatus(this.enemies, pacInput);

		this.pellets = pelletInput.map(p => new Pellet(new Position(p.x, p.y), p.value));

		// console.error(this.pacs);
		// console.error(this.enemies);
		// console.error(this.pellets);

		this.updateUniversalGrid();
	}

	updateUniversalGrid() {
		// for each pac, go through visibility area and update grid

		for (let pac of this.pacs) {
			for (let x = pac.pos.x; x < this.grid.width; x++) {
				const pos = new Position(x, pac.pos.y);
				if (this.grid.get(pos) === Entities.WALL) break;
				if (this.grid.get(pos) === Entities.EMPTY)
					this.universalGrid.set(pos, 0);
			}

			for (let x = pac.pos.x; x >= 0; x--) {
				const pos = new Position(x, pac.pos.y);
				if (this.grid.get(pos) === Entities.WALL) break;
				if (this.grid.get(pos) === Entities.EMPTY)
					this.universalGrid.set(pos, 0);
			}

			for (let y = pac.pos.y; y < this.grid.height; y++) {
				const pos = new Position(pac.pos.x, y);
				if (this.grid.get(pos) === Entities.WALL) break;
				if (this.grid.get(pos) === Entities.EMPTY)
					this.universalGrid.set(pos, 0);
			}

			for (let y = pac.pos.y; y >= 0; y--) {
				const pos = new Position(pac.pos.x, y);
				if (this.grid.get(pos) === Entities.WALL) break;
				if (this.grid.get(pos) === Entities.EMPTY)
					this.universalGrid.set(pos, 0);
			}
		}

		// this.universalGrid.debug();
	}
};

interface GameAction {
	type: 'MOVE' | 'SWITCH' | 'SPEED';
	into?: 'ROCK' | 'PAPER' | 'SCISSORS';
	pos?: Position;
};
class Solution {

	gameAi: GameAI;

	constructor(gameAi: GameAI) {
		this.gameAi = gameAi;
	}

	private getPossibleActionsForPac(pac: Pac) {
		const { pellets } = this.gameAi.gameState;

		const actions: GameAction[] = [];
		if (!pac.abilityCooldown) {
			actions.push({ type: 'SPEED' });
			actions.push({ type: 'SWITCH', into: 'ROCK' });
			actions.push({ type: 'SWITCH', into: 'PAPER' });
			actions.push({ type: 'SWITCH', into: 'SCISSORS' });
		}

		for (let i = 0; i < Math.min(pellets.length, 5); i++) {
			const index = Math.floor(Math.random() * pellets.length);
			actions.push({ type: 'MOVE', pos: pellets[index].pos });
		}

		return actions;
	}

	private generateMoves() {
		const { pacs } = this.gameAi.gameState;
		const alivePacs = pacs.filter(pac => !pac.isDead);
		for (let i = 0; i < pacs.length; i++) {
			const pac = pacs[i];
			const actions = this.getPossibleActionsForPac(pac);
			console.error(actions);
		}
	}

	private applyMovesAndEvalutate() {
	
	}

	randomizeAndEvaluate() {
		this.generateMoves();
	}
}

class GameAI {
	gameState: GameState;

	constructor(gameState: GameState) {
		this.gameState = gameState;
	}

	initPacTarget(pac: Pac, gridCopy: Grid<string>) {
		const { universalGrid } = this.gameState;
		const grid = gridCopy;
		const scores = new Grid<number>(grid.width, grid.height, -Infinity);
		const distances = new Grid<number>(grid.width, grid.height, Infinity);
		const visited = new Grid<boolean>(grid.width, grid.height, false);

		const setScore = (pos: Position, value: number) => {
			if (value > scores.get(pos)) {
				scores.set(pos, value);
			}
		};
		
		const setDistance = (pos: Position, value: number) => {
			// if (value > distances.get(pos)) {
				distances.set(pos, value);
			// }
		};

		const BFS = (initialPos: Position) => {
			let Q: { pos: Position, score: number, distance: number }[] = [];
			Q.push({ pos: new Position(initialPos.x, initialPos.y), score: 0, distance: 0 });

			while (Q.length) {
				let size = Q.length;
				while (size--) {
					const { pos, score, distance } = Q.shift();

					if (visited.get(pos)) continue;
					if (grid.get(pos) === Entities.WALL) continue;

					const neighbours = grid.getNeighbours(pos);

					// TODO: think more on this: adding -1 as a cost to walk
					let newVal: number = score + universalGrid.get(pos);
					setScore(pos, newVal);
					setDistance(pos, distance);

					for (const neighbour of neighbours) {
						Q.push({ pos: neighbour, score: newVal, distance: distance + 1 });
					}

					visited.set(pos, true);
				}
			};
		};

		BFS(pac.pos);


		const moves: { score: number, pos: Position }[] = [];
		// let bestScore: number = -1000, bestPos: Position = pac.pos;
		for (let y = 0; y < grid.height; y++) {
			for (let x = 0; x < grid.width; x++) {
				const pos = new Position(x, y);
				let distance = distances.get(pos), profit = scores.get(pos);
				const score = profit - (distance * 0.3);
				moves.push({ score, pos });
			}
		}

		const sortedMoves = moves.sort((m1, m2) => {
			return m1.score > m2.score ? -1 : 1;
		}).slice(0, 3);

		const randIndex = 0; //Math.floor(Math.random() * sortedMoves.length);
		const bestMove = sortedMoves[randIndex];

		universalGrid.debug();
		distances.debug();
		scores.debug();

		if (!pac.target) pac.target = { score: bestMove.score, pos: bestMove.pos };
		else if (bestMove.score - pac.target.score > 5) {
			pac.target = { score: bestMove.score, pos: bestMove.pos };
		}
	}

	playNextMove() {
		const { pacs, pellets } = this.gameState;
		const alivePacs = pacs.filter(pac => !pac.isDead);

		// reset targets if already reached
		alivePacs.forEach(pac => {
			if (pac.target && pac.pos.isEqual(pac.target.pos)) {
				pac.target = null;
			}
		});

		const gridCopy = this.gameState.grid.copy();

		let output = '';
		for (let i = 0; i < alivePacs.length; i++) {
			const pac = alivePacs[i];
			if (pac.isDead) continue;

			this.initPacTarget(pac, gridCopy);
			console.error(`${pac.id}: ${pac.target.pos.x} ${pac.target.pos.y}`);
			output += (`MOVE ${pac.id} ${pac.target.pos.x} ${pac.target.pos.y}`);
			if (i !== alivePacs.length - 1) {
				output += '|';
			}
		}

		console.log(output);
	}

};

function RunGame() {
	const gameState = new GameState();
	const gameAI = new GameAI(gameState);

	while (true) {
		gameState.initGameState();
		gameAI.playNextMove();
		// const sol = new Solution(gameAI);
		// sol.randomizeAndEvaluate();
	}

}

RunGame();
